pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java

```diff
@@ -290,63 +318,63 @@ public void run() {
-                instanceConfig.getFunctionDetails().getNamespace(),
-                instanceConfig.getFunctionDetails().getName()
-            );
-            StateStoreContext context = new StateStoreContextImpl();
-            store.init(context);
-
-            stateManager.registerStore(store);
-        }
-    }
-
-    private void processResult(Record srcRecord,
-                               CompletableFuture<JavaExecutionResult> result) throws Exception {
-        result.whenComplete((result1, throwable) -> {
-            if (throwable != null || result1.getUserException() != null) {
-                Throwable t = throwable != null ? throwable : result1.getUserException();
-                log.warn("Encountered exception when processing message {}",
-                        srcRecord, t);
-                stats.incrUserExceptions(t);
-                srcRecord.fail();
-            } else {
-                if (result1.getResult() != null) {
-                    sendOutputMessage(srcRecord, result1.getResult());
-                } else {
-                    if (instanceConfig.getFunctionDetails().getAutoAck()) {
-                        // the function doesn't produce any result or the user doesn't want the result.
-                        srcRecord.ack();
-                    }
-                }
-                // increment total successfully processed
-                stats.incrTotalProcessedSuccessfully();
+    @SuppressWarnings("rawtypes")
+	private void processResult(Record srcRecord, JavaExecutionResult result) throws Exception {
+    	
+    	if (result.getUserException() != null) {
+    		log.warn("Encountered exception when processing message {}",
+                    srcRecord, result.getUserException());
+            stats.incrUserExceptions(result.getUserException());
+            throw result.getUserException();
+    	} 
+    	
+    	if (result.getSystemException() != null) {
+    		log.warn("Encountered exception when processing message {}",
+                    srcRecord, result.getSystemException());
+            stats.incrSysExceptions(result.getSystemException());
+            throw result.getSystemException();
+    	}
+    	
+    	if (result.getResult() == null) {
+    		if (instanceConfig.getFunctionDetails().getAutoAck()) {
+                // the function doesn't produce any result or the user doesn't want the result.
+                srcRecord.ack();
             }
-        });
+    	} else { 
+    		
+    		try {
+				Object output = (result.getResult() instanceof CompletableFuture) ?
+					((CompletableFuture)result.getResult()).get() : result.getResult();
+				
+				sendOutputMessage(srcRecord, output);
+				
+				if (instanceConfig.getFunctionDetails().getAutoAck()) {
+					srcRecord.ack();
+				}
+				stats.incrTotalProcessedSuccessfully();
+				
+			} catch (InterruptedException | ExecutionException e) {
+				log.warn("Encountered exception when processing message {}",
```

Comment:
please handle InterruptedException

Reply:
@eolivelli Just to clarify, you want to call `Thread.currentThread().interrupt()` when we encounter an `InterruptedException`?  Why? and why only for that exception an not the `ExecutionException` ?

Reply:
because it is the standard Java practice.
when you catch a InterruptedException the "interrupted" flag is reset on the Thread and you have to set it again in order to let the information bubble up to the caller.

> why only for that exception an not the ExecutionException
because ExecutionException is not the special (and annoying) InterruptedException